mypy/checkmember.py

```diff
@@ -184,6 +184,74 @@ def analyze_member_access(name: str,
     return msg.has_no_attr(original_type, name, node)
 
 
+def analyze_descriptor_access(descriptor: Instance, owner: Instance, node: Context,
+                              is_lvalue: bool, is_class_access: bool,
+                              builtin_type: Callable[[str], Instance],
+                              msg: MessageBuilder,
+                              chk: 'mypy.checker.TypeChecker',
+                              original_type: Type = None) -> Optional[Type]:
+    """Analyse attribute access of a descriptor.
+
+    This is logically part of analyze_member_access and the arguments are
+    similar.
+    """
+    dunder_get = descriptor.type.get_method('__get__')
+    dunder_set = descriptor.type.get_method('__set__')
+    dunder_delete = descriptor.type.get_method('__delete__')
+    if not (dunder_get or dunder_set or dunder_delete):
+        return None
+
+    if is_lvalue:
+        if is_class_access:
+            # Fall back to setting the attribute as a normal class attribute
+            return None
+
+        if not dunder_set:
+            msg.fail("Can't set a data descriptor with no __set__ method", node)
+            return AnyType()
+
+        function = function_type(dunder_set, builtin_type('builtins.function'))
+        bound_method = bind_self(function, original_type)
+        typ = map_instance_to_supertype(descriptor, dunder_set.info)
+        dunder_set_type = expand_type_by_instance(bound_method, typ)
+
+        if not isinstance(dunder_set_type, CallableType):
+            msg.fail("__set__ must be callable", node)
+            return AnyType()
+
+        return dunder_set_type.arg_types[1]
```

Comment:
This assumes `__set__` has at least two arguments. But you don't check for that.

I think this is also approximately the point where you might need to check that the first argument matches the class in which the descriptor was encountered (typically 'A' in your tests).